source/logic.rst

```diff
@@ -0,0 +1,492 @@
+(If there is only one variable, you can use the singular ``intro``
+instead of ``intros``.)
+After the ``intros`` command,
+the goal is what it would have been at the start if we
+listed all the variables and hypotheses *before* the colon,
+as we did in the last section.
+In a moment, we will see why it is sometimes necessary to
+introduce variables and hypotheses after the proof begins.
+
+To help you prove the lemma, we will start you off:
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    -- BEGIN
+    lemma my_lemma : ∀ {x y ε : ℝ},
+      0 < ε → ε ≤ 1 → abs x < ε → abs y < ε → abs (x * y) < ε :=
+    begin
+      intros x y ε epos ele1 xlt ylt,
+      suffices h : abs x * abs y < 1 * ε,
+      { rw [←abs_mul, one_mul] at h,
+        apply h },
+      calc
+        abs x * abs y ≤ abs x * ε : sorry
+        ... < 1 * ε               : sorry
+    end
+    -- END
+
+We have introduced another new tactic here:
+``suffices`` works like ``have`` in reverse,
+asking you to prove the goal using the
+stated fact,
+and then leaving you the new goal of proving that fact.
+Finish the proof using the theorems ``mul_le_mul``,
+``abs_nonneg``, and ``mul_lt_mul_right``.
+Remember that you can find theorems like these using
+tab completion.
+Remember also that you can use ``.mp`` and ``.mpr``
+or ``.1`` and ``.2`` to extract the two directions
+of an if-and-only-if statement.
+
+Universal quantifiers are often hidden in definitions,
+and Lean will unfold definitions to expose them when necessary.
+For example, let's define two predicates,
+``fn_ub f a`` and ``fn_lb f a``,
+where ``f`` is a function from the real numbers to the real
+numbers and ``a`` is a real number.
+The first says that ``a`` is an upper bound on the
+values of ``f``,
+and the second says that ``a`` is a lower bound
+on the values of ``f``.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    -- BEGIN
+    def fn_ub (f : ℝ → ℝ) (a : ℝ) : Prop := ∀ x, f x ≤ a
+    def fn_lb (f : ℝ → ℝ) (a : ℝ) : Prop := ∀ x, a ≤ f x
+    -- END
+
+In the next example, remember that ``λ x, f x + g x`` is the
+function that maps ``x`` to ``f x + g x``.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    def fn_ub (f : ℝ → ℝ) (a : ℝ) : Prop := ∀ x, f x ≤ a
+    def fn_lb (f : ℝ → ℝ) (a : ℝ) : Prop := ∀ x, a ≤ f x
+
+    variables (f g : ℝ → ℝ) (a b : ℝ)
+
+    -- BEGIN
+    example (hfa : fn_ub f a) (hgb : fn_ub g b) :
+      fn_ub (λ x, f x + g x) (a + b) :=
+    begin
+      intro x,
+      dsimp,
+      apply add_le_add,
+      apply hfa,
+      apply hgb
+    end
+    -- END
+
+Applying ``intro`` to the predicate ``fn_ub (λ x, f x + g x) (a + b)``
+forces Lean to unfold the definition of ``fn_ub``
+and introduce ``x`` for the universal quantifier.
+The goal is then ``(λ (x : ℝ), f x + g x) x ≤ a + b``.
+But applying ``(λ x, f x + g x)`` to ``x`` should result in ``f x + g x``,
+and the ``dsimp`` command performs that simplification.
```

Comment:
Did you already explain `change`? You could mention it here, as it offers much more control.

Reply:
I think I used the syntactic equivalent `show` in the previous chapter. Can I stick with `show`?

Reply:
The issue with `show` is it works only with the goal, whereas you can `change ... at ...`.